src/ifcopenshell-python/ifcopenshell/validate.py

```diff
@@ -271,6 +273,51 @@ def get_entity_attributes(schema, entity):
     return entity_attrs
 
 
+def check(value, type, instance):
+    if value is None:
+        return
+
+    # @nb something can be a named type with rules and still be an aggregation.
+    # case in point IfcCompoundPlaneAngleMeasure. Therefore only unpack named
+    # type references from this point onwards.
+    while isinstance(
+        type,
+        (
+            ifcopenshell.ifcopenshell_wrapper.named_type,
+            ifcopenshell.ifcopenshell_wrapper.type_declaration,
+        ),
+    ):
+        type = type.declared_type()
+
+    if isinstance(value, (list, tuple)):
+        if isinstance(type, ifcopenshell.ifcopenshell_wrapper.aggregation_type):
+            ty = type.type_of_element()
+            for v in value:
+                res = yield from check(v, ty, instance=instance)
+                if res:
+                    yield res
+        else:
+            # Let's hope a schema validation error was reported for this case
+            pass
+
+    elif isinstance(value, ifcopenshell.entity_instance):
+        if isinstance(
+            value.is_a(),
+            ifcopenshell.ifcopenshell_wrapper.entity,
+        ):
+            # top level entity instances will be checked on their own
+            pass
+        else:
+            # unpack the type instance
+            yield from check(value[0], value.is_a(), instance=instance)
+
+        if value.is_a() == "IfcDateTime" and value[0]:
```

Comment:
This function still does the type check here, so can't be reused.

Rather yield the values and do the check outside of the function. That way the same function can also be used for checking uniqueness for example.